python/tvm/relay/tvm_rts_backend.py

```diff
@@ -0,0 +1,239 @@
+"""A compiler from Relay programs to TVM's graph runtime.
+"""
+import json
+from typing import Dict, Any, List, Tuple
+
+import attr
+
+from relay.frontend import get_env
+from . import ir
+from .tyck import get_checked_type
+from .opt import AbstractExprVisitor, compile_ops_to_module
+from ._make import Operator_is_generic
+
+
+@attr.s(auto_attribs=True)
+class NodeRef:
+    ident: int
+    index: int = 0
+    version: int = 0
+
+    def to_json(self) -> Any:
+        return [self.ident, self.index, self.version]
+
+
+@attr.s(auto_attribs=True)
+class Node():
+    name: str
+    attrs: Dict[str, Any]
+    is_output: bool
+
+    def to_json(self) -> Any:
+        raise Exception("Abstract method, please implement me.")
+
+
+@attr.s(auto_attribs=True)
+class InputNode(Node):
+    """An input node in the graph representation we lower to before NNVM's graph."""
+    is_output: bool = False
+
+    def to_json(self):
+        return {
+            "op": "null",
+            "name": self.name,
+            "inputs": []
+        }
+
+
+@attr.s(auto_attribs=True)
+class OpNode(Node):
+    """An operator node in the graph representation we lower to before NNVM's graph."""
+    op_name: str
+    inputs: List[NodeRef]
+    op_attrs: Dict[str, Any]
+    is_output: bool = False
+
+    def to_json(self) -> Any:
+        attrs = dict.copy(self.op_attrs)
+        # Extend ops with extra info.
+        attrs['func_name'] = self.op_name
+        # When do we flatten?
+        attrs['flatten_data'] = "0"
+        # Fix me!
+        attrs['num_inputs'] = str(len(self.inputs))
+        attrs['num_outputs'] = "1"
+
+        return {
+            "op": "tvm_op",
+            "name": self.name,
+            "attrs": attrs,
+            "inputs": self.inputs
+        }
+
+
+def from_tensor(typ: ir.TensorType) -> Tuple[str, List[int]]:
+    dtype = typ.dtype.dtype
+    shape = typ.shape
+    dims = []
+    for dim in shape.shapes:
+        dims.append(dim.value)
+    return dtype, dims
+
+
+class TVMRTSCompiler(AbstractExprVisitor[NodeRef]):
+    """The compiler from Relay to the TVM runtime system."""
+    nodes: List[Node]
+    id_map: Dict[ir.LocalId, NodeRef]
+
+    def __init__(self) -> None:
+        self.nodes = []
+        self.id_map = {}
+
+    def add_node(self, node: Node) -> NodeRef:
+        self.nodes.append(node)
+        ident = len(self.nodes) - 1
+        return NodeRef(ident)
+
+    def add_binding(self, ident: ir.LocalId, ref: NodeRef) -> None:
+        self.id_map[ident] = ref
+
+    def let_bind(self, ident: ir.LocalId, node: Node) -> NodeRef:
+        ref = self.add_node(node)
+        self.add_binding(ident, ref)
+        return ref
+
+    def get_node(self, ref: NodeRef) -> Node:
+        return self.nodes[ref.ident]
+
+    def lookup(self, ident: ir.LocalId) -> NodeRef:
+        return self.id_map[ident]
+
+    def compile(self, func: ir.Function) -> None:
```

Comment:
What is the advantages of keeping compile algorithm implemented in Python rather than in C++? I would say that keep it in C++ is better because it allows us to write DSL frontends for other languages more easily.

Reply:
Eventually we will port it to C++, the idea was to demonstrate some passes in C++ and Python, as the alpha of Relay progresses more and more will be moved into C++.